### PR TITLE
Test switching datastreams `index.mode`

### PR DESCRIPTION
**Pull request by salvatore-campagna**
Here we add a few integration tests where we try switching
`index.mode` considering both valid and invalid scenarios.

### CODE REVIEW

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/LogsDataStreamIT.java [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+        indexTimeSeriesDocuments(client(), randomIntBetween(10, 20), randomIntBetween(32, 64), dataStreamName);
+
+        updateComposableIndexTemplate(client(), "custom-composable-template", LOGS_OR_STANDARD_MAPPING, logsSettings, indexPatterns);
+        rolloverDataStream(dataStreamName);
+        indexLogOrStandardDocuments(client(), randomIntBetween(10, 20), randomIntBetween(32, 64), dataStreamName);
+
+        assertDataStreamBackingIndicesModes(dataStreamName, List.of(IndexMode.LOGS, IndexMode.TIME_SERIES, IndexMode.LOGS));
+    }
+
+    public void testInvalidIndexModeTimeSeriesSwitchWithoutROutingPath() throws IOException, ExecutionException, InterruptedException {
```
Comment (martijnvg) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/LogsDataStreamIT.java:227]: nit: ROuting -> Routing

**Review by martijnvg — APPROVED** [review]
I think going from logs to time series data stream is missing? Otherwise LGTM.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/LogsDataStreamIT.java [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.is;
+
+public class LogsDataStreamIT extends ESSingleNodeTestCase {
```
Comment (martijnvg) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/LogsDataStreamIT.java:57]: one other comment: maybe make this a rest level test (under `javaRestTest` directory)?

Comment (salvatore-campagna) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/LogsDataStreamIT.java:57]: Are you suggesting this should be a `ESTestCase`...in that case I would need to rewrite it...

Comment (martijnvg) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/LogsDataStreamIT.java:57]: Let like we chatted about. Let's not change this test, but add another rest java integration tests that at least tests going from normal data stream to logs data stream and logs data stream to normal data stream.

**Pull request merged**